src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -281,7 +412,37 @@ private Assembly ResolveUsingEvent(AssemblyName assemblyName)
             
             return assembly;
         }
-        
+
+        /// <summary>
+        /// This method is called back by the native code after Unloading has been initiated
+        /// This method is called indirectly by the finalizer of the LoaderAllocator
+        /// </summary>
+        private void OnUnloading()
+        {
+            lock (unloadLock)
+            {
+                if (state == InternalState.Unloading)
+                {
+                    state = InternalState.Unloaded;
+                }
+                else
+                {
+                    // Otherwise we didn't have time to be called here and we will be 
+                    // called via the finalizer
+                    return;
+                }
+            }
+            var unloading = Unloading;
+            // TODO: should we enclose this with a try catch?
```

Comment:
Good point, yea I think this needs to have a try/catch but per delegate invocation, such that all registered handlers are called regardless of one throwing an exception. I have filed a separate issue for this, dotnet/corefx#14747.